Update ACCESS-OM2 run instructions (#653)
ACCESS-NRI/access-hive.org.au: docs/models/run-a-model/run-access-om.md

    <terminal-line>Created and checked out new branch: expt</terminal-line>
    <terminal-line>laboratory path:  /scratch/.../access-om2</terminal-line>
    <terminal-line>binary path:  /scratch/.../access-om2/bin</terminal-line>
    <terminal-line>input path:  /scratch/.../access-om2/input</terminal-line>
    <terminal-line>work path:  /scratch/.../access-om2/work</terminal-line>
    <terminal-line>archive path:  /scratch/.../access-om2/archive</terminal-line>
    <terminal-line>Updated metadata. Experiment UUID: daeee7ff-07e4-4f93-823b-cb7c6e4bdb6e</terminal-line>
    <terminal-line>Added archive symlink to /scratch/.../access-om2/archive/1deg_jra55_ryf-expt-daeee7ff</terminal-line>
    <terminal-line>To change directory to control directory run:</terminal-line>
    <terminal-line>    cd 1deg_jra55_ryf</terminal-line>
</terminal-window>

In the example above the `payu clone` command clones the 1° repeat-year JRA55 configuration (` -B release-1deg_jra55_ryf`) 
to a new experiment branch (`-b expt`) to a directory named `1deg_jra55_ryf`: 

    mkdir -p ~/access-om

>     cd ~/access-om2

Reviewer: We've got a typo/inconsistency between dir names here: `access-om` & `access-om2`

Author: Made it consistent throughout document to be `~/access-om2`

## Running an {{ model }} configuration

{{ model }} configurations run on [_Gadi_](https://opus.nci.org.au/display/Help/How+to+connect+to+a+project) through a [PBS Job][PBS Jobs] submission managed by `payu`.

The general layout of a `payu`-supported model run consists of two main directories:

- The **laboratory** directory, where all the model components reside. For {{ model }}, it is typically `/scratch/$PROJECT/$USER/access-om2`.

-  The **control directory** contains the model configuration and serves as the execution directory for running the model (in this example, the cloned directory `~/access-om/1deg_jra55_ryf`).

This separates the small text configuration files from the larger binary outputs and inputs. In this way, the _control_ directory can be in the `$HOME` directory (as it is the only filesystem actively backed-up on _Gadi_). The quotas for `$HOME` are low and strict, which limits what can be stored there, so it is not suitable for larger files.

Furthermore, this separation allows multiple self-resubmitting experiments that share common executables and input data to be run simultaneously.

To setup the _laboratory_ directory, from the _control_ directory run:

>     payu init

Reviewer: From the view of a new user, does the init command _need_ to be run?

Author: Can you confirm that it isn't necessary? I don't believe it is, but would be good to have that tested. Note that if the laboratory already exists to do a proper test the laboratory directories would have to be deleted, or the test run using a different project so there was not an existing laboratory.

Author: I have [confirmed `payu init` is not required](https://github.com/payu-org/payu/issues/428#issuecomment-2050853660). Removed.

    <terminal-line>Created and checked out new branch: expt</terminal-line>
    <terminal-line>laboratory path:  /scratch/.../access-om2</terminal-line>
    <terminal-line>binary path:  /scratch/.../access-om2/bin</terminal-line>
    <terminal-line>input path:  /scratch/.../access-om2/input</terminal-line>
    <terminal-line>work path:  /scratch/.../access-om2/work</terminal-line>
    <terminal-line>archive path:  /scratch/.../access-om2/archive</terminal-line>
    <terminal-line>Updated metadata. Experiment UUID: daeee7ff-07e4-4f93-823b-cb7c6e4bdb6e</terminal-line>
    <terminal-line>Added archive symlink to /scratch/.../access-om2/archive/1deg_jra55_ryf-expt-daeee7ff</terminal-line>
    <terminal-line>To change directory to control directory run:</terminal-line>
    <terminal-line>    cd 1deg_jra55_ryf</terminal-line>
</terminal-window>

In the example above the `payu clone` command clones the 1° repeat-year JRA55 configuration (` -B release-1deg_jra55_ryf`) 
to a new experiment branch (`-b expt`) to a directory named `1deg_jra55_ryf`: 

>      mkdir -p ~/access-om

Reviewer 2: ```suggestion
    mkdir -p ~/access-om
```

>      cd ~/access-om2

Reviewer 2: ```suggestion
    cd ~/access-om2
```

Which will kill the current job without waiting for it to complete. If you have used the `-n` option, but decide you no longer wish to keep running after the current process completes, you can create a file called `stop_run` in the _control_ directory, and this will prevent `payu` from submitting another job.

### Error and output log files

While the model is running, _payu_ saves the model standard output and standard error in the `access-om2.out` and `access-om2.err` log files in the _control_ directory. You can examine the contents of these files to check on the status of a run as it progresses.

At the end of a successful run these log files are archived. If they remain in the _control_ directory after the PBS job for a run has completed this is an indication the run has failed.

### Did the model run correctly?

To determine if a model has run correctly it must first be established that it has finished. The `qstat` commands [above](#monitor-access-om2-runs) and the [presence of PBS log files](#pbs-output-files) should be used to determine if the PBS job has ended.

If the model did not run to completion correctly the following will still be in the `control` directory:

>     work

Reviewer 2: ```suggestion
    work/
```

    access-om2.err
    access-om2.out

This is because `payu` will only run the `archive` step when the model runs without error.

## {{ model }} outputs

At the end of a successful model run, output files, restart files and log files are moved from the work directory to the archive directory. A symbolic link in the control directory to a directory in the _laboratory_ (`/scratch/$PROJECT/$USER/access-om2/archive`) is provided for convenience.

If a model run is unsuccessful the `work` directory is left untouched to facilitate "run forensics" to determine the cause of the model failure.

Outputs and restarts are stored in subfolders within the `archive`, subdivided for each run of the model.

Output folders are `outputXXX` and restart folders `restartXXX`, where _XXX_ is the run number starting from `000`.

Model components are separated into subdirectories within the output and restart directories.

<terminal-window>

> <terminal-line data="input">cd ~/access-om/1deg_jra55_yaf<</terminal-line>

Reviewer 2: ```suggestion
<terminal-line data="input">cd ~/access-om/1deg_jra55_ryf<</terminal-line>
```

If you are unsure whether {{ model }} is the right choice for your experiment, take a look at the overview of [ACCESS Models](/models).

> ### Conventions

Reviewer 2: Is this better in a note block ?

Author: Good suggestion. Thanks. Done.

### Model-specific prerequisites

**Join the _vk83_ and _qv56_ projects at NCI** 

To join these projects request membership on the respective [vk83](https://my.nci.org.au/mancini/project/vk83/join) and [qv56](https://my.nci.org.au/mancini/project/qv56/join) NCI project pages.

For more information on how to join specific NCI projects, please refer to [How to connect to a project](https://opus.nci.org.au/display/Help/How+to+connect+to+a+project").

**Payu**

[_Payu_](https://github.com/payu-org/payu) is a workflow management tool for running numerical models in supercomputing environments for which there is extensive [documentation](https://payu.readthedocs.io/en/latest/). 

_Payu_ on _Gadi_ is available through a dedicated `conda` environment in the _vk83_ project.

After joining the _vk83_ project load the _payu_ environment:

    module use /g/data/vk83/modules
    module load payu

To check that `payu` is available, run:

    payu --version

<terminal-window>
<terminal-line data="input">payu --version</terminal-line>
<terminal-line lineDelay="1000">1.1.1</terminal-line>
</terminal-window>

> 

Author: Sorry, yes this is a known issue I should have flagged. The deployment was auto-updating the version string. It does now, but was only finally fixed after the last deployment. I have on my TO-DO list to push a new `payu` tag and update `payu` which will mean it will report a proper version. I am am waiting for a few `payu` fixes before doing this.

**Note:** `payu` version >=1.1 is required

----------------------------------------------------------------------------------------

## Get {{ model }} configuration

> All {{ model }} configurations are available from the [ACCESS-OM2 configs] GitHub repository.  ACCESS-NRI has adapted model configurations from those originally developed by [COSIMA]. 

Author: Good point. I'll add a "released" specifier in there. Might have to add a sentence explaining what a released configuration actually *is*. 

There aren't all those configurations yet, but there will be when this is merged.

 
There are global configurations for three resolutions: 1°, 0.25°, 0.1°. For each resolution there are two options of atmospheric forcing: Repeat Year (RYF) and Interannual (IAF). Each configuration also has a biogeochemical (BGC) configuration if this is required. Note the BGC experiments are slower and so consume more resources, both compute time and generally also disk space. 

Each configuration is stored as a separate specially named branch in the [ACCESS-OM2 configs] GitHub repository. They are organised like this for administrative convenience: they are easier to manage and keep grouped for testing. Anyone using a configuration is advised to just clone a single branch and not attempt to keep this structure. The [ACCESS-OM2 configs] repo has more information about the available experiments and the naming scheme of the branches.

> The first step is to decide which configuration is required from the twelve available. For example, if the 1° horizontal resolution configuration with repeat-year JRA55 forcing (without bgc) is the required configuration then the [`release-1deg_jra55_ryf`](https://github.com/ACCESS-NRI/access-om2-configs/tree/release-1deg_jra55_ryf) branch is the correct configuration.

Author: It's a good point. Thanks.

The next step is to clone this branch to a location on _Gadi_:

    payu clone -b expt -B release-1deg_jra55_ryf https://github.com/COSIMA/1deg_jra55_ryf.git 1deg_jra55_ryf

In the example above the `payu clone` command clones the 1° repeat-year JRA55 configuration (` -B release-1deg_jra55_ryf`) 
to a new experiment branch (`-b expt`) to a directory named `1deg_jra55_ryf`. 

!!! note

    `payu` uses branches to differentiate between different experiments the same git repository. So it is recommended to always change the branch name you clone into to something meaningful for the planned experiment. See the [`payu` tutorial](https://forum.access-hive.org.au/t/access-om2-payu-tutorial/1750#select-experiment-12) for more information.

<terminal-window>
    <terminal-line data="input">mkdir -p ~/access-om</terminal-line>

>     <terminal-line data="input">cd ~/access-om</terminal-line>

Author: Yep. I can see the value in the terminal thingo, but it doesn't allow for easy copy n' paste of a command, which the code blocks do.

I'll rearrange, but the clone after the terminal and just highlight the command that was run to allow people to copy it out of a code block. See if that works better.

>     <terminal-line data="input">cd ~/access-om</terminal-line>

Reviewer 2: Maybe ... i think it's just "different"

>     <terminal-line data="input">cd ~/access-om</terminal-line>

Reviewer 2: Should we just add a comment that thats the purpose of having both - black shows demos, grey is for copy  /paste

>     <terminal-line data="input">cd ~/access-om</terminal-line>

Author: Good idea. Have done so.

----------------------------------------------------------------------------------------

## Running an {{ model }} configuration

{{ model }} configurations run on [_Gadi_](https://opus.nci.org.au/display/Help/How+to+connect+to+a+project) through a [PBS Job][PBS Jobs] submission managed by `payu`.

The general layout of a `payu`-supported model run consists of two main directories:

- The **laboratory** directory, where all the model components reside. For {{ model }}, it is typically `/scratch/$PROJECT/$USER/access-om2`.

- The **control** directory, where the model configuration resides and from where the model is run (in this example, the cloned directory `~/access-om/1deg_jra55_iaf`).

This separates the small text configuration files from the larger binary outputs and inputs. In this way, the _control_ directory can be in the `$HOME` directory (as it is the only filesystem actively backed-up on _Gadi_). The quotas for `$HOME` are low and strict, which limits what can be stored there, so it is not suitable for larger files.

Furthermore, this separation allows multiple self-resubmitting experiments that share common executables and input data to be run simultaneously.

> To setup the _laboratory_ directory, from the _control_ directory run the:

Author: Thanks for all the feedback. To make this easier and more precise [code suggestions](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request) are a great feature.

> To setup the _laboratory_ directory, from the _control_ directory run the:

Reviewer 2: Ah yes! I kept selecting the word and clicking reply and being surprised when i got an empy comment !

>     payu init

Reviewer 2: Ping?

>     payu init

Author: It actually isn't currently, though that has brought this up as a potential area that needs fixing.

https://github.com/payu-org/payu/issues/428

So it should stay until the issue is fixed.

This creates the _laboratory_ directory, together with relevant subdirectories, depending on the configuration. 
The main subdirectories of interest are: 

- `work` &rarr; a temporary directory where the model is run. It gets created as part of a run and then removed after the run succeeds.
- `archive` &rarr; the directory where output is stored after each run.

<terminal-window>

> <terminal-line data="input">cd ~/access-om/1deg_jra55_yaf</terminal-line>

Author: Agreed.

<terminal-line data="input" directory="~/access-om/1deg_jra55_iaf">payu init</terminal-line>
<terminal-line>laboratory path:  /scratch/$PROJECT/$USER/access-om2</terminal-line>
<terminal-line>binary path:  /scratch/$PROJECT/$USER/access-om2/bin</terminal-line>
<terminal-line>input path:  /scratch/$PROJECT/$USER/access-om2/input</terminal-line>
<terminal-line>work path:  /scratch/$PROJECT/$USER/access-om2/work</terminal-line>
<terminal-line>archive path:  /scratch/$PROJECT/$USER/access-om2/archive</terminal-line>
</terminal-window>

### Payu setup (optional)

As a first step, from within the _control_ directory, it is good practice to run:

>     payu setup

Author: It *is* really useful, but yeah point taken. I'll move it to the end and call it trouble-shooting or debugging.

Each configuration is stored as a separate specially named branch in the [ACCESS-OM2 configs] GitHub repository.
Anyone using a configuration is advised to clone only a single branch and not attempt to keep this structure. 
The [ACCESS-OM2 configs] repo has more information about the available experiments and the naming scheme of the branches.

The first step is to choose a configuration from those available. For example, if the 1° horizontal resolution configuration with repeat-year JRA55 forcing (without bgc) is the required configuration then the [`release-1deg_jra55_ryf`](https://github.com/ACCESS-NRI/access-om2-configs/tree/release-1deg_jra55_ryf) branch is the correct configuration.

The next step is to clone this branch to a location on _Gadi_:

<terminal-window>
    <terminal-line data="input">mkdir -p ~/access-om</terminal-line>
    <terminal-line data="input">cd ~/access-om</terminal-line>

>     <terminal-line data="input" directory="~/access-om">payu clone -b expt -B release-1deg_jra55_ryf https://github.com/COSIMA/1deg_jra55_ryf.git 1deg_jra55_ryf</terminal-line>

Reviewer 2: ```suggestion
    <terminal-line data="input" directory="~/access-om">payu clone -b expt -B release-1deg_jra55_ryf https://github.com/ACCESS-NRI/access-om2-configs.git 1deg_jra55_ryf</terminal-line>
```

    <terminal-line>Created and checked out new branch: expt</terminal-line>
    <terminal-line>laboratory path:  /scratch/.../access-om2</terminal-line>
    <terminal-line>binary path:  /scratch/.../access-om2/bin</terminal-line>
    <terminal-line>input path:  /scratch/.../access-om2/input</terminal-line>
    <terminal-line>work path:  /scratch/.../access-om2/work</terminal-line>
    <terminal-line>archive path:  /scratch/.../access-om2/archive</terminal-line>
    <terminal-line>Updated metadata. Experiment UUID: daeee7ff-07e4-4f93-823b-cb7c6e4bdb6e</terminal-line>
    <terminal-line>Added archive symlink to /scratch/.../access-om2/archive/1deg_jra55_ryf-expt-daeee7ff</terminal-line>
    <terminal-line>To change directory to control directory run:</terminal-line>
    <terminal-line>    cd 1deg_jra55_ryf</terminal-line>
</terminal-window>

In the example above the `payu clone` command clones the 1° repeat-year JRA55 configuration (` -B release-1deg_jra55_ryf`) 
to a new experiment branch (`-b expt`) to a directory named `1deg_jra55_ryf`: 

> 

Reviewer 2: ```suggestion
     mkdir -p ~/access-om
     cd ~/access-om2
```

<terminal-window>
<terminal-line data="input">cd ~/access-om/1deg_jra55_yaf<</terminal-line>
<terminal-line data="input" directory="~/access-om/1deg_jra55_yaf<">ls</terminal-line>
<terminal-line class="ls-output-format">output000 pbs_logs restart000</terminal-line>
</terminal-window>

### Model Live Diagnostics

ACCESS-NRI developed the [Model Live Diagnostics](/model_evaluation/#model-live-diagnostics) framework to check, monitor, visualise, and evaluate model behaviour and progress of ACCESS models currently running on _Gadi_.

For a complete documentation on how to use this framework, check the [Model Diagnostics documentation](https://med-live-diagnostics.readthedocs.io/en/latest/index.html).

### Trouble-shooting

If `payu` doesn't run correctly for some reason a good first step step, from within the _control_ directory, is to run:

    payu setup -f

This will prepare the model run: create the ephemeral `work` directory based on the experiment configuration, generate manifests and report some useful information to the user, such as the location of the _laboratory_ where the `work` and `archive` directories are located.
<terminal-window>

> <terminal-line data="input">payu setup</terminal-line>

Reviewer 2: ```suggestion
<terminal-line data="input">payu setup -f</terminal-line>
```

Author: Thanks! Glad you've to see you've joined team force.

Reviewer 2: I haven't. But i like consistency !

<terminal-line>Writing manifests/exe.yaml</terminal-line>
</terminal-window>

!!! note

    This step can be skipped as it is also included in the run command. However, running it explicitly helps to check for errors and ensure executable and restart directories are accessible.

### Run configuration

> To run an {{ model }} configuration: 

Author: The jinga template will replace `{{ model }}` with ACCESS-OM2, so the sentence becomes:

> To run an ACCESS-OM2 configuration

To be honest it could be removed completely, and just be

> To run a configuration.

I will make that change.

>     payu run -f

Reviewer 3: I suggest avoiding the use of `-f` in the guide. But it could be mentioned as an option, similar to what was done below.

If a previous run fails and leaves data in work, running payu run would trigger a 'pause', indicating that there was an issue with the previous run. This would guide the user to address any errors encountered in the former run. Otherwise, careless users might assume that all their runs were successful, only to find no data in archive when attempting to post-process.

Author: Ok. I bow to the collective wisdom.  Will remove them all.

> This will submit a single job to the queue with a run length of `restart_period`.  `restart_period` is defined in the `accessom2.nml` file in the _control_ directory.

Author: It is fairly side-effect free. I personally think `payu run` shouldn't care if `work` exists, but I lost that argument. If people use `payu run -f` as a matter of course it will avoid a lot of confused people.

Happy to concede if enough people disagree with me ... ;)

> This will submit a single job to the queue with a run length of `restart_period`.  `restart_period` is defined in the `accessom2.nml` file in the _control_ directory.

Reviewer 2: Ok i haven't thought about the specific useage lot . Just using `-f` without thought gives me alarm bells!

    <terminal-line>&lt;job-ID&gt;.gadi-pbs</terminal-line>
</terminal-window>

### Run configuration multiple times

> If you want to run an {{ model }} configuration multiple times automatically use the option `-n`:

Author: See above. Will remove explicit mention of the model.

>     payu run -f -n <number-of-runs>

Author: See above. My opinion has changed since then ... :)

This will run `number-of-runs` times with a total length of `restart_period * number-of-runs`.

> For example, to run a configuration for a total of 50 years with `restart_period = 5, 0, 0` (5 years), the `number-of-runs` should be set to `10`:

Author: Yep. Will slightly re-word and add a note linking to that section.

>     payu run -f -n 10

Author: I think we've established by now that this my jam.

## Monitor {{ model }} runs

> Currently, there is no specific tool to monitor {{ model }} runs. However `payu run` reports the PBS `job-ID`, e.g. `110020843.gadi-pbs`, as the last line to the terminal. `qstat` can be used to query the status of the job, e.g.

Author: Agreed.

    <terminal-line linedelay=0>&lt;110021035&gt;.gadi-pbs&nbsp;&nbsp;&nbsp;1deg_jra55_ryf&nbsp;&nbsp;&nbsp;&lt;$USER&gt;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&lt;time&gt;&nbsp;R&nbsp;normal-exec</terminal-line>
    <terminal-line linedelay=0>&lt;000000000&gt;.gadi-pbs&nbsp;&nbsp;&nbsp;&lt;other-job-name&gt;&nbsp;&lt;$USER&gt;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&lt;time&gt;&nbsp;R&nbsp;normal-exec</terminal-line>
    <terminal-line linedelay=0>&lt;000000000&gt;.gadi-pbs&nbsp;&nbsp;&nbsp;&lt;other-job-name&gt;&nbsp;&lt;$USER&gt;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&lt;time&gt;&nbsp;R&nbsp;normal-exec</terminal-line>
</terminal-window>
 The default name of the your job is the name of the _payu_ _control_ directory, in this example `1deg_jra55_ryf`. This can be changed by altering the `jobname` in [set in the `config.yaml`](#modify-pbs-resources).

_S_ indicates the status of your run, where:

- _Q_ &rarr; Job waiting in the queue to start
- _R_ &rarr; Job running
- _E_ &rarr; Job ending

> If there are no jobs listed with your `jobname` (or if no job is listed), your run either successfully completed or was terminated due to an error.

Reviewer 2: Maybe move section ### PBS output files to just here ?

> If there are no jobs listed with your `jobname` (or if no job is listed), your run either successfully completed or was terminated due to an error.

Author: > review the pbs log

Not sure this is the right place for that. I think a dedicated article about PBS and debugging might be a better option and link to it. 

It's not just that I'm trying to get out of work, but I'm concerned about scope-creep for this, and maintainability.

> If there are no jobs listed with your `jobname` (or if no job is listed), your run either successfully completed or was terminated due to an error.

Author: I have moved the PBS files section up though.

> If there are no jobs listed with your `jobname` (or if no job is listed), your run either successfully completed or was terminated due to an error.

Reviewer 2: I guess it seemed like there isnt a way to positively confirm the model ran ... Are we saying the creation of an `archive` directory is the sign of success? Otherwise you need to look in the log. (Reviewing it isnt hard ... you open and check it doesn't say Error?)

> If there are no jobs listed with your `jobname` (or if no job is listed), your run either successfully completed or was terminated due to an error.

Author: Oh right. If the model log files `access-om2.err` and `access-om2.out` and the `work` directory are still there then the model hasn't run correctly.

It says this, but maybe not explicitly enough. I'll try and improve this.

> If there are no jobs listed with your `jobname` (or if no job is listed), your run either successfully completed or was terminated due to an error.

Author: Have added a section "Did the model run correctly".

Now realise I have used configuration and model fairly much interchangeably. We need a style guide!

### Stop a run

If you want to manually terminate a run, you can do so by executing:

    qdel job-ID

Which will kill the current job without waiting for it to complete. If you have used the `-n` option, but decide you no longer wish to keep running after the current process completes, you can create a file called `stop_run` in the _control_ directory, and this will prevent `payu` from submitting another job.

### Error and output log files

While the model is running, _payu_ saves the model standard output and standard error in the `access-om2.out` and `access-om2.err` log files in the _control_ directory. You can examine the contents of these files to check on the status of a run as it progresses.

> At the end of a successful run these log files are archived. If they remain in the _control_ directory after the PBS job for a run has completed this is an indication the run has failed.

Reviewer 3: For first time users, I suggest to be more explicit. We are spooning this to users in a straightforward way. :)

```suggestion
At the end of a successful run these log files are archived to the `archive` directory and will not be found in the _control_ directory. If they remain in the _control_ directory after the PBS job for a run has completed this is an indication the run has failed.
```

### PBS output files

> When the model completes PBS writes the standard outout and error streams to two files into the _control_ directory: `jobname.o<job-ID>` and `jobname.e<job-ID>` respectively. This is terminal output that isn't otherwise redirected into model log files.

Reviewer 2: You can archive these files using `payu sweep` which moves them to the `archive` directory.

## {{ model }} outputs

> At the end of a successful model run, output files (and restart files) are moved (archived) from the `work` directory to the `archive` directory which is a symbolic link to a directory in the _laboratory_ (`/scratch/$PROJECT/$USER/access-om2/archive`). The model log files are also moved to the `archive` directory. 

Reviewer 2: At the end of a successful model run, output files, restart files and log files are moved from the `work` directory to the `archive` directory.  A symbolic link in the _control_ directory to a directory in the _laboratory_ (`/scratch/$PROJECT/$USER/access-om2/archive`) is provided for convenience.

For a complete documentation on how to use this framework, check the [Model Diagnostics documentation](https://med-live-diagnostics.readthedocs.io/en/latest/index.html).

---

## Modifying an {{ model }} configuration

Once you are comfortable that you can clone and run an existing configuration it is often the case that you will want to modify the configuration depending on your science goals.

This section describes the model configuration and how to modify it. 

Modifications can be to the way the model is run by `payu`, or can change the way specific model components are configured, or the coupling between them. Sometimes changes are required to both, if the model component changes require a change to the resources needed for the model to complete.

The `config.yaml` file located in the _control_ directory, is the _Master Configuration_ file, which controls the general model configuration. It contains several parts, some of which it is more likely will need modification, and others are rarely changed without significant understanding of how the model is configured.

### Change run length

> One of the most common changes is to how long the model runs. For example it is common when debugging changes to a model to want to reduce the run length to reduce the amount of resources wasted and get faster feedback on changes.

Reviewer 3: ```suggestion
One of the most common changes is to adjust the duration of the model run. For example when debugging changes to a model, it is common to reduce the run length to minimise resource consumption and return faster feedback on changes.
```

> To change the run length, edit the `restart_period` field in the `&date_manager_nml` section of the `~/access-om/1deg_jra55_iaf/accessom2.nml` file:

Reviewer 3: iaf->ryf

This can either be a number, which retains every _nth_ numbered restart, or a pandas style date-time frequency alias. For example to preserve the ability to restart the model every 50 model run years:
```yaml
restart_freq: 50Y
```
The most recent sequential restarts are retained, and only deleted only after a permanently archived restart has been produced.

See the [documentation for more detail](https://payu.readthedocs.io/en/latest/config.html#model).

### Other rarely modified configuration options

#### Model configuration

This tells `payu` which driver to use for the main model configuration (`access-om2`) and the location of all inputs that are common to all the component models, or submodels.

```yaml
name: common
model: access-om2
input: /g/data/ik11/inputs/access-om2/input_20201102/common_1deg_jra55
```
The `name` field here is not actually used for the configuration run so you can safely ignore it.

#### Submodels

{{ model }} is a coupled model deploying multiple submodels (i.e. [model components]).

This section specifies the submodels and configuration options required to execute the model correctly.

> Each submodel contains additional configuration options that are read in when the submodel is running. These options are specified in the subfolder of the _control_ directory, whose name matches the submodel's _name_ (e.g., configuration options for the `ocean` submodel are in the `~/access-om/1deg_jra55_iaf/ocean` directory).

Reviewer 3: iaf->ryf

This should not be changed as it is an essential part of the provenance of an experiment. `payu` updates the manifest files for every run, and relies on `runlog` to save this information in the `git` history so there is a record of all inputs, restarts and executables used in an experiment.

- **Stack size**

```yaml
stacksize: unlimited
```

`stacksize` is the maximum size (in KiB) of the per-thread resources allocated for each process. This is often set to `unlimited` as some Fortran programs can use a large amount of stack memory, and stack memory errors can be difficult to diagnose.

- **mpirun arguments**

```yaml
mpirun: --mca io ompio --mca io_ompio_num_aggregators 1
```

Line to append mpirun arguments to the `mpirun` call of the model.

- **_qsub_ arguments**

```yaml
qsub_flags: -W umask=027
```

Additional command line arguments for the _qsub_ command. In this case the `umask` setting makes sure PBS output files are produced with group readable permissions. 

- **Environment variables**

```yaml
env:
UCX_LOG_LEVEL: 'error'
```

Sets environment variables within the PBS job.

> 

Reviewer 2: I kind of think the four items above are quite niche (stack size, mpirun arguments, qsub argumnts, environment variables), and unlikely to be relevant in an intro the running the model guide?

Author: I agree, which is why I moved them down into a Miscellaneous section. 

I'll aggregate them, and hide them by default.

- **Platform-specific defaults**

```yaml
platform: 
nodesize: 48
```

Set platform-specific default parameters.  In this case it sets the default number of cpus per node to 48.

- **User scripts**

```yaml
userscripts:
    error: resub.sh
    run: rm -f resubmit.count
```

A dictionary to run scripts or subcommands at various stages of a _payu_ submission.

`error` gets called if the model does not run correctly and returns an error code.  `run` gets called after the model execution, but prior to model output archive

To find out more about other configuration settings for the `config.yaml` file, check out [how to configure your experiment with _payu_](https://payu.readthedocs.io/en/latest/config.html").

### Edit a single {{ model }} component configuration

> Each of the [model components] contains configuration options specific to that model that are read in when the model component is running. These options are typically useful to modify the physics used in the model or the input data.

Reviewer 2: the physics used in the model, the input data or the variables saved in the history output.

Author: I've used "diagnostic output". I think "history files" is quite specific to some models, and isn't widely used in other areas. Happy to be told otherwise if I'm wrong.

Reviewer 2: DId you mean to make a change here?

to me: diagnostic output means output (often in a log file) used to diagnose if the model is faulty, history output is the variable output you might use for climate analysis

Author: > DId you mean to make a change here?

Yes. I did, but hadn't pushed, thanks.

> to me: diagnostic output means output (often in a log file) used to diagnose if the model is faulty, history output is the variable output you might use for climate analysis

The ocean community use the term "diagnostic output" for output data.

I've gone with

"These options are typically useful to modify the physics used in the model, the input data or the model variables saved in the output files."

> These configuration options are are specified in files in a subfolder of the _control_ directory, named the same as the submodel's name in the `config.yaml` `submodel` section (e.g., configuration options for the _ocean_ submodel are in the `~/access-om/1deg_jra55_iaf/ocean` directory).

Reviewer 3: iaf->ryf

#### Miscellaneous

There rest of the configuration settings should never need changing: `stacksize`, `mpirun`, `qsub_flags` and `env`.

??? Show configuration details

    ```yaml
    stacksize: unlimited
    mpirun: --mca io ompio --mca io_ompio_num_aggregators 1
    qsub_flags: -W umask=027
    env:
    UCX_LOG_LEVEL: 'error'
    ```

To find out more about other configuration settings for the `config.yaml` file, refer to [how to configure your experiment with `payu`](https://payu.readthedocs.io/en/latest/config.html").

### Edit a single {{ model }} component configuration

Each of the [model components] contains configuration options specific to that model that are read in when the model component is running. These options are typically useful to modify the physics used in the model, the input data or the variables saved in the diagnostic output. 
These configuration options are are specified in files in a subfolder of the _control_ directory, named the same as the submodel's name in the `config.yaml` `submodel` section (e.g., configuration options for the _ocean_ submodel are in the `~/access-om/1deg_jra55_iaf/ocean` directory).

> To modify these options please refer to the User Guide of each individual model component.

Reviewer 2: Do we have a neat way to link to these? I think not?

Author: Yeah that's a bit of a cop-out eh? Off the top of my head we don't provide any such thing. We should. Don't know if we can in any reasonable time-frame, but should make it a TO-DO.

Reviewer 2: Yeah - i think its hard though. Its not obvious where cice 5 documentation is. e.g. its not in the cice readthedocs history, and then it was modified a fair bit by COSIMA. (as an example)

Author: It is a lot tougher for legacy models we're taking over for sure. But it reasonable that we make some effort to collect together what is out there.

Author: I have made issues in the model component repros to try and get this happening.

---

# References

- [COSIMA]

> - [Kiss et al. (2020)](http://doi.org/10.5194/gmd-13-401-2020")

Reviewer 2: ```suggestion
- [Kiss et al. (2020)](http://doi.org/10.5194/gmd-13-401-2020)
```

Reviewer 2: This one :). Do this one :)

> - https://github.com/COSIMA/access-om2/

Reviewer 2: ```suggestion
- https://github.com/access-nri/access-om2/
```

    ```yaml
    stacksize: unlimited
    mpirun: --mca io ompio --mca io_ompio_num_aggregators 1
    qsub_flags: -W umask=027
    env:
    UCX_LOG_LEVEL: 'error'
    ```

To find out more about other configuration settings for the `config.yaml` file, refer to [how to configure your experiment with `payu`](https://payu.readthedocs.io/en/latest/config.html").

### Edit a single {{ model }} component configuration

Each of the [model components] contains configuration options specific to that model that are read in when the model component is running. These options are typically useful to modify the physics used in the model, the input data or the variables saved in the diagnostic output. 
These configuration options are are specified in files in a subfolder of the _control_ directory, named the same as the submodel's name in the `config.yaml` `submodel` section (e.g., configuration options for the _ocean_ submodel are in the `~/access-om/1deg_jra55_iaf/ocean` directory).
To modify these options please refer to the User Guide of each individual model component.

---

# References

- [COSIMA]
- [Kiss et al. (2020)](http://doi.org/10.5194/gmd-13-401-2020)

> - https://github.com/access-nri/access-om2/

Reviewer 2: I am not sure why https://github.com/access-nri/access-om2/ is not getting formatted as a link in the build preview

Author: Fixed.
